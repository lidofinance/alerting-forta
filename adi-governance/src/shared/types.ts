export interface IProxyContractData {
  name: string
  shortABI: string
}
