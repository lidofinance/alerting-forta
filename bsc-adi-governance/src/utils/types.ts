import { FindingSeverity, FindingType, ethers } from 'forta-agent'
import { Log } from '@ethersproject/abstract-provider'

export type EventOfNotice = {
  address: string
  event: string
  alertId: string
  name: string
  description: CallableFunction
  severity: FindingSeverity
  type: FindingType
}

export type TransactionDto = {
  logs: Log[]
  to: string | null
  block: {
    timestamp: number
    number: number
  }
}

export type LogDescription = ethers.utils.LogDescription
