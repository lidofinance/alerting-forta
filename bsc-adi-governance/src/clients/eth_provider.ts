import { TransactionResponse } from '@ethersproject/abstract-provider'
import { ethers } from 'forta-agent'
import * as E from 'fp-ts/Either'
import { retryAsync } from 'ts-retry'
import { NetworkError } from '../utils/finding.helpers'

const DELAY_IN_500MS = 500
const ATTEMPTS_5 = 5

export class ETHProvider {
  private readonly jsonRpcProvider: ethers.providers.JsonRpcProvider

  constructor(jsonRpcProvider: ethers.providers.JsonRpcProvider) {
    this.jsonRpcProvider = jsonRpcProvider
  }

  public async getStartedBlockForApp(argv: string[]): Promise<E.Either<Error, number>> {
    let latestBlockNumber: number = -1

    if (argv.includes('--block')) {
      latestBlockNumber = parseInt(argv[4])
    } else if (argv.includes('--range')) {
      latestBlockNumber = parseInt(argv[4].slice(0, argv[4].indexOf('.')))
    } else if (argv.includes('--tx')) {
      const txHash = argv[4]
      const tx = await this.getTransaction(txHash)
      if (E.isLeft(tx)) {
        return E.left(tx.left)
      }

      if (tx.right.blockNumber !== undefined) {
        latestBlockNumber = tx.right.blockNumber
      }
    }
    if (latestBlockNumber == -1) {
      try {
        latestBlockNumber = await this.jsonRpcProvider.getBlockNumber()
      } catch (e) {
        return E.left(new NetworkError(e, `Could not fetch latest block number`))
      }
    }

    return E.right(latestBlockNumber)
  }

  public async getTransaction(txHash: string): Promise<E.Either<Error, TransactionResponse>> {
    try {
      const out = await retryAsync<TransactionResponse>(
        async (): Promise<TransactionResponse> => {
          const tx = await this.jsonRpcProvider.getTransaction(txHash)

          if (!tx) {
            throw new Error(`Can't find transaction ${txHash}`)
          }

          if (tx.blockNumber === undefined) {
            throw new Error(`Transaction ${txHash} was not yet included into block`)
          }

          return tx
        },
        { delay: DELAY_IN_500MS, maxTry: ATTEMPTS_5 },
      )

      return E.right(out)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not call getTransaction`))
    }
  }
}
