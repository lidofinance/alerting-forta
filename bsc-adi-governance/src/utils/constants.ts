export const CROSS_CHAIN_EXECUTOR_ADDRESS = '0x8e5175d17f74d1d512de59b2f5d5a5d8177a123d'
export const CROSS_CHAIN_CONTROLLER_ADDRESS = '0x40c4464fca8cacd550c33b39d674fc257966022f'
