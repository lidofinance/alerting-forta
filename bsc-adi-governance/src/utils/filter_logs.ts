import { ethers } from 'forta-agent'
import { LogDescription, TransactionDto } from './types'

export function filterLogs(txEvent: TransactionDto, eventSignature: string, eventAddress: string): LogDescription[] {
  const iface = new ethers.utils.Interface([eventSignature])
  const filteredLogs: LogDescription[] = []

  for (const log of txEvent.logs) {
    if (log.address.toLowerCase() !== eventAddress.toLowerCase()) {
      continue
    }

    try {
      const logDescription = iface.parseLog(log)
      filteredLogs.push(logDescription)
    } catch (e) {
      // Only one from eventsOfNotice could be correct
      // Others - skipping
    }
  }

  return filteredLogs
}
