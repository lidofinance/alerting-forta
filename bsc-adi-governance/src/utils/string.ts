export function bscscanAddress(address: string): string {
  return `[${address}](https://bscscan.com/address/${address})`
}
