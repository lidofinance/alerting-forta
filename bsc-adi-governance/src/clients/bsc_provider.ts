import { TransactionResponse } from '@ethersproject/abstract-provider'
import { ethers } from 'forta-agent'
import { either as E } from 'fp-ts'
import { retryAsync } from 'ts-retry'
import { ICRossChainControllerClient } from '../services/cross-chain-controller/CrossChainController.srv'
import { NetworkError } from '../utils/errors'
import { BaseAdapter__factory, CrossChainController } from '../generated'
import { CROSS_CHAIN_CONTROLLER_ADDRESS } from '../utils/constants'

const DELAY_IN_500MS = 500
const ATTEMPTS_5 = 5
const MAINNET_CHAIN_ID = 1

export class BSCProvider implements ICRossChainControllerClient {
  private jsonRpcProvider: ethers.providers.JsonRpcProvider
  private crossChainControllerRunner: CrossChainController

  constructor(jsonRpcProvider: ethers.providers.JsonRpcProvider, crossChainControllerRunner: CrossChainController) {
    this.jsonRpcProvider = jsonRpcProvider
    this.crossChainControllerRunner = crossChainControllerRunner
  }

  public async getTransaction(txHash: string): Promise<E.Either<Error, TransactionResponse>> {
    try {
      const out = await retryAsync<TransactionResponse>(
        async (): Promise<TransactionResponse> => {
          const tx = await this.jsonRpcProvider.getTransaction(txHash)

          if (!tx) {
            throw new Error(`Can't find transaction ${txHash}`)
          }

          if (tx.blockNumber === undefined) {
            throw new Error(`Transaction ${txHash} was not yet included into block`)
          }

          return tx
        },
        { delay: DELAY_IN_500MS, maxTry: ATTEMPTS_5 },
      )

      return E.right(out)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not call getTransaction`))
    }
  }

  public async getBlockNumber(): Promise<E.Either<Error, number>> {
    try {
      const latestBlockNumber = await this.jsonRpcProvider.getBlockNumber()
      return E.right(latestBlockNumber)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not fetch latest block number`))
    }
  }

  public async getBridgeAdapters(): Promise<E.Either<Error, string[]>> {
    try {
      const bridgeAdapters = await this.crossChainControllerRunner.getReceiverBridgeAdaptersByChain(MAINNET_CHAIN_ID)
      return E.right(bridgeAdapters)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not get bridge adapters list from CrossChainController contract`))
    }
  }

  public async getBridgeAdaptersNamesMap(): Promise<E.Either<Error, Map<string, string>>> {
    try {
      const bridgeAdapters = await this.getBridgeAdapters()
      if (E.isLeft(bridgeAdapters)) {
        return E.left(bridgeAdapters.left)
      }
      const result = new Map<string, string>()

      for (const adapterAddress of bridgeAdapters.right) {
        const bridgeAdapterContract = BaseAdapter__factory.connect(adapterAddress, this.jsonRpcProvider)
        try {
          const adapterName = await bridgeAdapterContract.adapterName()
          result.set(adapterAddress, adapterName)
        } catch (error) {
          return E.left(new NetworkError(error, `Could not get adapter name from ${adapterAddress}`))
        }
      }

      return E.right(result)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not get bridge adapters list from CrossChainController contract`))
    }
  }

  public async getEnvelopeStateByIds(envelopeIds: string[]): Promise<E.Either<Error, Map<string, number>>> {
    try {
      const out = new Map<string, number>()
      await Promise.all(
        envelopeIds.map(async (envelopeId) => {
          const status = await this.crossChainControllerRunner['getEnvelopeState(bytes32)'](envelopeId)
          out.set(envelopeId, status)
        }),
      )
      return E.right(out)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not get bridge adapters list from CrossChainController contract`))
    }
  }

  public async getReceivedEnvelopeIds(fromBlock: number, toBlock: number): Promise<E.Either<Error, string[]>> {
    try {
      const results = await retryAsync(
        async () => {
          const keccakEvent = ethers.utils.keccak256(
            ethers.utils.toUtf8Bytes('TransactionReceived(bytes32,bytes32,uint256,(uint256,bytes),address,uint8)'),
          )
          const address = CROSS_CHAIN_CONTROLLER_ADDRESS
          return this.jsonRpcProvider.getLogs({
            fromBlock: `0x${fromBlock.toString(16)}`,
            toBlock: `0x${toBlock.toString(16)}`,
            address,
            topics: [keccakEvent, null, '0x0000000000000000000000000000000000000000000000000000000000000001', null],
          })
        },
        { delay: DELAY_IN_500MS, maxTry: ATTEMPTS_5 },
      )
      const iface = new ethers.utils.Interface([
        'event TransactionReceived(bytes32 transactionId, bytes32 indexed envelopeId, uint256 indexed originChainId, (uint256,bytes), address indexed bridgeAdapter, uint8 confirmations)',
      ])
      const out = results.map((result) => iface.parseLog(result))
      const res: string[] = out.map((out) => out.args.envelopeId)
      return E.right(res)
    } catch (e) {
      return E.left(new NetworkError(e, `Could not call jsonRpcProvider.getStonksOrderParams`))
    }
  }
}
