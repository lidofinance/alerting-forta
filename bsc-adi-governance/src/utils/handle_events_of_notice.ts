import { Finding, FindingType } from 'forta-agent'
import { EventOfNotice, TransactionDto } from './types'
import { filterLogs } from './filter_logs'

export function handleEventsOfNotice(txEvent: TransactionDto, eventsOfNotice: EventOfNotice[]) {
  const findings: Finding[] = []

  for (const eventInfo of eventsOfNotice) {
    const filteredLogs = filterLogs(txEvent, eventInfo.event, eventInfo.address)
    for (const filteredLog of filteredLogs) {
      findings.push(
        Finding.fromObject({
          name: eventInfo.name,
          description: eventInfo.description(filteredLog.args),
          alertId: eventInfo.alertId,
          severity: eventInfo.severity,
          type: FindingType.Info,
          metadata: { args: String(filteredLog.args) },
        }),
      )
    }
  }

  return findings
}
