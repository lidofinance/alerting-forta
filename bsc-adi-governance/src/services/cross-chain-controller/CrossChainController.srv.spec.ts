import { TransactionDto } from '../../utils/types'
import * as Winston from 'winston'
import { BlockEvent, ethers, Finding, getEthersProvider } from 'forta-agent'

import { CrossChainControllerSrv } from './CrossChainController.srv'
import { TransactionEvent } from 'forta-agent/dist/sdk/transaction.event'
import { BSCProvider } from '../../clients/bsc_provider'
import { CrossChainController__factory } from '../../generated'
import { CROSS_CHAIN_CONTROLLER_ADDRESS } from '../../utils/constants'
import { getFortaConfig } from 'forta-agent/dist/sdk/utils'

const TEST_TIMEOUT = 120_000 // ms

describe('CrossChainController Srv functional tests', () => {
  const chainId = 56
  const ethProvider = new ethers.providers.JsonRpcProvider(getFortaConfig().jsonRpcUrl, chainId)

  const ethersProvider = getEthersProvider()

  const crossChainControllerRunner = CrossChainController__factory.connect(
    CROSS_CHAIN_CONTROLLER_ADDRESS,
    ethersProvider,
  )

  const bscClient = new BSCProvider(ethersProvider, crossChainControllerRunner)

  const logger: Winston.Logger = Winston.createLogger({
    format: Winston.format.simple(),
    transports: [new Winston.transports.Console()],
  })

  const service = new CrossChainControllerSrv(logger, bscClient, CROSS_CHAIN_CONTROLLER_ADDRESS)

  let runTransaction: (txHash: string, initBlock: number) => Promise<Finding[]>
  let runBlock: (blockNumber: number, initBlock: number) => Promise<Finding[]>

  beforeAll(() => {
    jest.spyOn(console, 'log').mockImplementation(() => {})
    jest.spyOn(Date, 'now').mockImplementation(() => new Date('2024-01-01').getTime())

    runTransaction = async (txHash: string, initBlock: number) => {
      const trx = await ethProvider.getTransaction(txHash)
      const receipt = await trx.wait()
      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
      }
      await service.initialize(initBlock)
      return service.handleTransaction(transactionDto as TransactionEvent)
    }

    runBlock = async (blockNumber, initBlock) => {
      if (initBlock > Number(blockNumber)) {
        throw new Error('Wrong initial block or argument order')
      }
      const initErr = await service.initialize(initBlock)
      if (initErr instanceof Error) {
        throw initErr
      }
      const BlockEventHack = { blockNumber } as BlockEvent
      return service.handleBlock(BlockEventHack)
    }
  })

  afterAll(() => {
    jest.resetAllMocks()
  })

  it(
    'should find ConfirmationsUpdated',
    async () => {
      const findings = await runTransaction(
        '0x1042618d67813e40d9cae2d4e4f72924963c72455140ca921ab29eaff5aaed55',
        39776560,
      )
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should find TransactionReceived',
    async () => {
      const findings = await runTransaction(
        '0x81b925698152319f40e667c12d0da2370562f4f0d9f9c166ce7bcddbc5f0dcb3',
        39920557,
      )
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should find 4 ReceiverBridgeAdaptersUpdated',
    async () => {
      const findings = await runTransaction(
        '0x86a1f2ae752fec82b647b6dfecedcf702ba83161b92b2dd9228e3a3dde3f2462',
        39776551,
      )
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should find GuardianUpdated and AdminChanged and Upgraded(proxy)',
    async () => {
      const findings = await runTransaction(
        '0xda1572eb6360116ea0fbd9f898b1413823ba2ab4bb1cbb5e150cce42c68a687c',
        39775722,
      )
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should find 2 receive event but message is delivered',
    async () => {
      const findings = await runBlock(39921800, 39921799)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )
})
