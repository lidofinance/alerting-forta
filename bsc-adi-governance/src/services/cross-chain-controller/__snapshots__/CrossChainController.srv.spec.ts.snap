// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrossChainController Srv functional tests should find 2 receive event but message is delivered 1`] = `[]`;

exports[`CrossChainController Srv functional tests should find 4 ReceiverBridgeAdaptersUpdated 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "BSC-ADI-BRIDGE-SET-UPDATED",
    "description": "Allowed bridges set updated - added 0x15AD245133568c2498c7dA0cf2204A03b0e9b98A",
    "labels": [],
    "metadata": {
      "args": "0x15AD245133568c2498c7dA0cf2204A03b0e9b98A,true,1",
    },
    "name": "üö® BSC a.DI: Allowed bridges set updated",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "BSC-ADI-BRIDGE-SET-UPDATED",
    "description": "Allowed bridges set updated - added 0xc934433f4c433Cf80DE6fB65fd70C7a650D8a408",
    "labels": [],
    "metadata": {
      "args": "0xc934433f4c433Cf80DE6fB65fd70C7a650D8a408,true,1",
    },
    "name": "üö® BSC a.DI: Allowed bridges set updated",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "BSC-ADI-BRIDGE-SET-UPDATED",
    "description": "Allowed bridges set updated - added 0xCd867B440c726461e5fAbe8d3a050b2f8701C230",
    "labels": [],
    "metadata": {
      "args": "0xCd867B440c726461e5fAbe8d3a050b2f8701C230,true,1",
    },
    "name": "üö® BSC a.DI: Allowed bridges set updated",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "BSC-ADI-BRIDGE-SET-UPDATED",
    "description": "Allowed bridges set updated - added 0xBb1E43408BbF2C767Ff3Bd5bBC34E183CC1Ef119",
    "labels": [],
    "metadata": {
      "args": "0xBb1E43408BbF2C767Ff3Bd5bBC34E183CC1Ef119,true,1",
    },
    "name": "üö® BSC a.DI: Allowed bridges set updated",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;

exports[`CrossChainController Srv functional tests should find ConfirmationsUpdated 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "BSC-ADI-CONFIRMATIONS-UPDATED",
    "description": "Allowed bridges quorum updated to 3",
    "labels": [],
    "metadata": {
      "args": "3,1",
    },
    "name": "üö® BSC a.DI: Allowed Bridges quorum updated",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;

exports[`CrossChainController Srv functional tests should find GuardianUpdated and AdminChanged and Upgraded(proxy) 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "PROXY-ADMIN-CHANGED",
    "description": "Proxy admin for bnbCrossChainControllerAddress (0x40c4464fca8cacd550c33b39d674fc257966022f) was changed from 0x0000000000000000000000000000000000000000 to 0x29E6817db339795766244B96aEf5Dc534a98518d
(detected by event)",
    "labels": [],
    "metadata": {
      "args": "0x0000000000000000000000000000000000000000,0x29E6817db339795766244B96aEf5Dc534a98518d",
    },
    "name": "üö®üö®üö® BSC a.DI: Proxy admin changed",
    "protocol": "ethereum",
    "severity": 5,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "PROXY-UPGRADED",
    "description": "Proxy for bnbCrossChainControllerAddress (0x40c4464fca8cacd550c33b39d674fc257966022f) was updated to 0xB7Ba81dd07885ae7BFD18452B36D3404d7EDD8Ee
(detected by event)",
    "labels": [],
    "metadata": {
      "args": "0xB7Ba81dd07885ae7BFD18452B36D3404d7EDD8Ee",
    },
    "name": "üö®üö®üö® BSC a.DI: Proxy upgraded",
    "protocol": "ethereum",
    "severity": 5,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
  Finding {
    "addresses": [],
    "alertId": "BSC-ADI-GUARDIAN-UPDATED",
    "description": "Guardian was updated from 0x0000000000000000000000000000000000000000 to 0x0000000000000000000000000000000000000000",
    "labels": [],
    "metadata": {
      "args": "0x0000000000000000000000000000000000000000,0x0000000000000000000000000000000000000000",
    },
    "name": "üö®üö®üö® BSC a.DI: Guardians updated",
    "protocol": "ethereum",
    "severity": 5,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;

exports[`CrossChainController Srv functional tests should find TransactionReceived 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "BSC-ADI-MESSAGE-RECEIVED",
    "description": "CrossChainController received a message from 0xCd867B440c726461e5fAbe8d3a050b2f8701C230 (Hyperlane adapter)",
    "labels": [],
    "metadata": {
      "args": "0x9a3408ef1a0ec12a5081af3a1ea3b669dd0b2caecf617ec81647c71a5ccced13,0x0c39ef7274a750de961d1ee0e2083ecccacef2c4c5bd3499317f3f2177e43470,1,0,0x00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000184d39300f2fa4419d04998e9c58cb5de586d879000000000000000000000000250227c30e3b26525f292f63756e5eb3c425427e0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000003800000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000038000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000000000000000002400000000000000000000000000000000000000000000000000000000000000320000000000000000000000000000000000000000000000000000000000000000200000000000000000000000040c4464fca8cacd550c33b39d674fc257966022f00000000000000000000000029e6817db339795766244b96aef5dc534a98518d000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000001a7472616e736665724f776e657273686970286164647265737329000000000000000000000000000000000000000000000000000000000000000000000000001a7472616e736665724f776e65727368697028616464726573732900000000000000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000200000000000000000000000008e5175d17f74d1d512de59b2f5d5a5d8177a123d00000000000000000000000000000000000000000000000000000000000000200000000000000000000000008e5175d17f74d1d512de59b2f5d5a5d8177a123d000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000,0xCd867B440c726461e5fAbe8d3a050b2f8701C230,1",
    },
    "name": "‚ÑπÔ∏è BSC a.DI: Message received by the Hyperlane adapter",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;
